File: cloudflare_worker/worker.py
import json
import httpx
from html.parser import HTMLParser
from patreon_tier_alerter.src.alerter import check_tiers

# Global cache used across invocations
alerted_tiers_cache = {}

async def scrape_patreon_page_async(creator_url: str, user_agent: str, client: httpx.AsyncClient):
    """Asynchronously fetch and parse a Patreon creator page."""
    headers = {"User-Agent": user_agent}
    try:
        resp = await client.get(creator_url, headers=headers, timeout=10)
        resp.raise_for_status()
    except httpx.HTTPError as e:
        print(f"Error fetching URL {creator_url}: {e}")
        return None

    class TierParser(HTMLParser):
        def __init__(self):
            super().__init__()
            self.tiers = []
            self.current = None
            self.in_btn_div = False

        def handle_starttag(self, tag, attrs):
            attrs = dict(attrs)
            if tag == "a" and attrs.get("data-tag") == "patron-checkout-continue-button":
                aria_label = attrs.get("aria-label", "")
                if not aria_label:
                    return
                tier_name = " ".join(aria_label.split()[:-1])
                disabled = attrs.get("aria-disabled") == "true"
                self.current = {"name": tier_name, "disabled": disabled, "button": ""}
            elif self.current and tag == "div" and attrs.get("class") == "cm-oHFIQB":
                self.in_btn_div = True

        def handle_data(self, data):
            if self.current and self.in_btn_div:
                self.current["button"] += data.strip()

        def handle_endtag(self, tag):
            if tag == "div" and self.in_btn_div:
                self.in_btn_div = False
            elif tag == "a" and self.current:
                text = self.current.get("button", "")
                if self.current["disabled"] or text == "Sold Out":
                    status = "sold_out"
                elif text == "Join":
                    status = "available"
                else:
                    status = "unknown"
                if self.current["name"]:
                    self.tiers.append({"name": self.current["name"], "status": status})
                self.current = None

    parser = TierParser()
    try:
        parser.feed(resp.text)
    except Exception as e:
        print(f"Error parsing HTML from {creator_url}: {e}")
        return None

    return parser.tiers

async def send_sms_alerts_async(alerts_to_send: list, env: dict, client: httpx.AsyncClient):
    """Send SMS alerts using a generic HTTP API."""
    sms_url = env.get("SMS_API_URL")
    sms_token = env.get("SMS_API_TOKEN")
    phone = env.get("RECIPIENT_PHONE_NUMBER")
    if not alerts_to_send or not all([sms_url, sms_token, phone]):
        return
    for alert in alerts_to_send:
        message = (
            f"Patreon Alert: Tier '{alert['tier_name']}' for creator '{alert['creator_name']}' "
            f"is now available! Check at: {alert['url']}"
        )
        payload = {"to": phone, "token": sms_token, "message": message}
        try:
            await client.post(sms_url, json=payload, timeout=10)
        except Exception as e:
            print(f"Error sending SMS: {e}")

async def main(request, env):
    """Entry point for the Cloudflare Worker."""
    config_text = env.get("CONFIG_JSON")
    if not config_text and hasattr(env, "KV_CONFIG"):
        config_text = await env.KV_CONFIG.get("config", "text")
    if not config_text:
        return {"error": "missing configuration"}
    config = json.loads(config_text)
    creators = config.get("creators", [])
    user_agent = config.get("user_agent", "Mozilla/5.0")
    alerts = []
    async with httpx.AsyncClient() as client:
        for creator in creators:
            scraped = await scrape_patreon_page_async(creator.get("url"), user_agent, client)
            if scraped is None:
                continue
            new_alerts = check_tiers(scraped, creator, alerted_tiers_cache)
            alerts.extend(new_alerts)
        await send_sms_alerts_async(alerts, env, client)
    return {"alerts": alerts}


async def on_fetch(request, env, ctx):
    """Cloudflare Workers fetch handler."""
    return await main(request, env)

File: cloudflare_worker/src/worker.js
// Cloudflare Worker: Patreon Tier Alerter (KV-based)
// - Config and alert cache are stored in KV
// - SMS sent via Twilio
// - Designed for Cron Triggers

import { DOMParser } from "linkedom";

export default {
  async scheduled(event, env, ctx) {
    // 1. Load config from KV
    const configRaw = await env.PATREON_CONFIG.get('config');
    if (!configRaw) {
      console.error('No config found in KV!');
      return;
    }
    const config = JSON.parse(configRaw);

    // 2. For each creator, scrape and check tiers
    for (const creator of config.creators) {
      const tiers = await scrapePatreonTiers(creator.url, config.user_agent);
      if (!tiers) continue;
      const alerts = await checkTiers(tiers, creator, env.PATREON_ALERT_CACHE);
      if (alerts.length > 0) {
        await sendAlerts(alerts, config.sms_settings, env);
      }
    }
  },
};

// --- Patreon Scraper ---
async function scrapePatreonTiers(url, userAgent) {
  try {
    const resp = await fetch(url, { headers: { 'User-Agent': userAgent } });
    if (!resp.ok) return null;
    const html = await resp.text();
    // Use DOMParser to extract tier info
    const doc = new DOMParser().parseFromString(html, 'text/html');
    // Adjust selectors as needed for Patreon
    const tierEls = doc.querySelectorAll('a[data-tag="patron-checkout-continue-button"]');
    const tiers = [];
    tierEls.forEach(el => {
      const ariaLabel = el.getAttribute('aria-label') || '';
      const tierName = ariaLabel.split(' ').slice(0, -1).join(' ');
      const disabled = el.getAttribute('aria-disabled') === 'true';
      let buttonText = '';
      // Patreon may have a div inside the button for text
      el.childNodes.forEach(child => {
        if (child.nodeType === 3) buttonText += child.textContent.trim();
        if (child.nodeType === 1) buttonText += child.textContent.trim();
      });
      let status = 'unknown';
      if (disabled || buttonText === 'Sold Out') status = 'sold_out';
      else if (buttonText === 'Join') status = 'available';
      tiers.push({ name: tierName, status });
    });
    return tiers;
  } catch (e) {
    console.error('Error scraping Patreon:', e);
    return null;
  }
}

// --- Tier Checker ---
async function checkTiers(scrapedTiers, creator, alertCacheKV) {
  const newlyAvailable = [];
  const scrapedMap = {};
  scrapedTiers.forEach(t => scrapedMap[t.name.toLowerCase()] = t);
  for (const tierName of creator.tiers_to_watch) {
    const cacheKey = `${creator.name}_${tierName}`;
    const found = scrapedMap[tierName.toLowerCase()];
    if (found && found.status === 'available') {
      const alerted = await alertCacheKV.get(cacheKey);
      if (!alerted) {
        newlyAvailable.push({ creator_name: creator.name, tier_name: tierName, url: creator.url });
        await alertCacheKV.put(cacheKey, '1');
      }
    } else {
      // Reset alert cache if not available
      await alertCacheKV.delete(cacheKey);
    }
  }
  return newlyAvailable;
}

// --- Alert Sender (Twilio) ---
async function sendAlerts(alerts, smsConfig, env) {
  if (!alerts.length) return;
  // Log alerts
  alerts.forEach(alert => {
    console.log(`ALERT: Tier "${alert.tier_name}" for creator "${alert.creator_name}" is now available! ${alert.url}`);
  });
  // Send SMS via Twilio
  if (smsConfig.provider === 'twilio') {
    const { twilio_account_sid, twilio_auth_token, twilio_from_number, recipient_phone_number } = smsConfig;
    if (!twilio_account_sid || !twilio_auth_token || !twilio_from_number || !recipient_phone_number) {
      console.error('Twilio config incomplete.');
      return;
    }
    for (const alert of alerts) {
      const body = `Patreon Alert: Tier '${alert.tier_name}' for creator '${alert.creator_name}' is now available! ${alert.url}`;
      const creds = btoa(`${twilio_account_sid}:${twilio_auth_token}`);
      await fetch(`https://api.twilio.com/2010-04-01/Accounts/${twilio_account_sid}/Messages.json`, {
        method: 'POST',
        headers: {
          'Authorization': `Basic ${creds}`,
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: new URLSearchParams({
          From: twilio_from_number,
          To: recipient_phone_number,
          Body: body,
        }),
      });
    }
  }
}

// --- Notes ---
// - Set up two KV namespaces: PATREON_CONFIG (for config), PATREON_ALERT_CACHE (for alert state)
// - Store your config JSON in PATREON_CONFIG with key 'config'
// - Add your Twilio credentials to the config in KV
// - Use Wrangler to bind KV namespaces in wrangler.toml
// - Set up a Cron Trigger in wrangler.toml for periodic checks 